File: python/src/asqav/_schema.py
"""Lightweight, dependency-free context-schema validator (criterion 328).

Schema spec: a dict mapping field names to a field descriptor.

Field descriptor keys:
    type     - one of "string" | "number" | "boolean" | "object" | "array"
    required - bool, defaults to False

Example::

    schema = {
        "user_id": {"type": "string", "required": True},
        "score":   {"type": "number"},
        "tags":    {"type": "array"},
    }

Pass to ``Agent.sign(..., context_schema=schema)``. The SDK validates the
context dict against the schema and raises ``AsqavValidationError`` on
mismatch before any network call. On success the context is returned with
keys in stable sorted order so audit trails are consistent and queryable.

You may instead pass a callable as ``context_schema``:

    def my_validator(ctx: dict) -> None:
        import jsonschema
        jsonschema.validate(ctx, MY_FULL_JSON_SCHEMA)

    agent.sign("action", ctx, context_schema=my_validator)

A callable receives the (possibly None-coerced-to-{}) context dict and must
raise on invalid input. The built-in path needs no new runtime dependency.
"""

from __future__ import annotations

from typing import Any

# Docs URL stamped on AsqavValidationError instances from this module.
_SCHEMA_DOCS_URL = "https://asqav.com/docs/structured-receipts"

# Maps schema "type" strings to the Python runtime type(s) they check against.
_TYPE_MAP: dict[str, type | tuple[type, ...]] = {
    "string": str,
    "number": (int, float),
    "boolean": bool,
    "object": dict,
    "array": list,
}

_VALID_TYPE_NAMES = frozenset(_TYPE_MAP)


def validate_context_schema(
    context: dict[str, Any] | None,
    schema: dict[str, Any],
) -> None:
    """Validate *context* against the built-in *schema* descriptor.

    Raises ``AsqavValidationError`` on the first offending field so the
    developer sees one clear message rather than a list dump.

    Args:
        context: The context dict passed to ``Agent.sign()``. ``None`` is
            treated as an empty dict (no user keys supplied).
        schema:  A ``{field_name: {"type": ..., "required": ...}}`` mapping.

    Raises:
        AsqavValidationError: When a required field is absent or a value
            does not match its declared type.
        ValueError: When the schema itself uses an unknown type name; this
            is a programmer error, not a runtime data error.
    """
    # Lazy import to avoid circular dependency with client.py.
    from .client import AsqavValidationError

    ctx = context or {}

    for field, descriptor in schema.items():
        required = descriptor.get("required", False)
        type_name = descriptor.get("type")

        if type_name is not None and type_name not in _VALID_TYPE_NAMES:
            raise ValueError(
                f"context_schema_error: unknown type '{type_name}' for field '{field}'; "
                f"must be one of {sorted(_VALID_TYPE_NAMES)}"
            )

        if field not in ctx:
            if required:
                raise AsqavValidationError(
                    f"context_schema_error: field '{field}' is required but missing",
                    docs_url=_SCHEMA_DOCS_URL,
                )
            continue

        if type_name is None:
            continue

        value = ctx[field]
        expected = _TYPE_MAP[type_name]

        # bool is a subclass of int; reject it from "number" explicitly so
        # {"active": True} does not silently pass a "number" field.
        if type_name == "number" and isinstance(value, bool):
            raise AsqavValidationError(
                f"context_schema_error: field '{field}' expected type number, got bool",
                docs_url=_SCHEMA_DOCS_URL,
            )

        if not isinstance(value, expected):
            actual = type(value).__name__
            raise AsqavValidationError(
                f"context_schema_error: field '{field}' expected type {type_name}, got {actual}",
                docs_url=_SCHEMA_DOCS_URL,
            )


def normalize_context(context: dict[str, Any] | None) -> dict[str, Any] | None:
    """Return *context* with keys in stable sorted order.

    Stable key order means two sign() calls with identical data but
    different insertion orders produce the same signed bytes, making
    receipts consistent and queryable.

    Returns ``None`` unchanged; an empty dict becomes ``{}``.
    """
    if context is None:
        return None
    return dict(sorted(context.items()))

File: typescript/src/schema.ts
/** Opt-in context-schema validator for agent.sign() (criterion 328). */

/** Valid type names for built-in field descriptors. */
export type ContextSchemaFieldType =
  | "string"
  | "number"
  | "boolean"
  | "object"
  | "array";

/** Per-field descriptor in a ContextSchema. */
export interface ContextSchemaField {
  /** Expected JavaScript type of the value. */
  type: ContextSchemaFieldType;
  /** When true the field must be present. Defaults to false. */
  required?: boolean;
}

/** Built-in schema: a record of field names to ContextSchemaField descriptors. */
export type ContextSchema = Record<string, ContextSchemaField>;

/** Docs URL stamped on ValidationError instances. */
const SCHEMA_DOCS_URL = "https://asqav.com/docs/structured-receipts";

/** Validate context against the schema; throws ValidationError on first bad field.
 * null/undefined context treated as empty ({}). */
export function validateContextSchema(
  context: Record<string, unknown> | null | undefined,
  schema: ContextSchema,
): void {
  const ctx = context ?? {};

  for (const [field, descriptor] of Object.entries(schema)) {
    const { type: typeName, required = false } = descriptor;

    const validTypes: readonly string[] = [
      "string",
      "number",
      "boolean",
      "object",
      "array",
    ];
    if (!validTypes.includes(typeName)) {
      throw new Error(
        `context_schema_error: unknown type '${typeName}' for field '${field}'; ` +
          `must be one of ${validTypes.join(", ")}`,
      );
    }

    if (!(field in ctx)) {
      if (required) {
        throw new ValidationError(
          `context_schema_error: field '${field}' is required but missing`,
          SCHEMA_DOCS_URL,
        );
      }
      continue;
    }

    const value = ctx[field];

    if (typeName === "number") {
      // bool is rejected: typeof true === "boolean", not "number", but guard explicitly.
      if (typeof value === "boolean") {
        throw new ValidationError(
          `context_schema_error: field '${field}' expected type number, got boolean`,
          SCHEMA_DOCS_URL,
        );
      }
      if (typeof value !== "number") {
        throw new ValidationError(
          `context_schema_error: field '${field}' expected type number, got ${typeof value}`,
          SCHEMA_DOCS_URL,
        );
      }
      continue;
    }

    if (typeName === "array") {
      if (!Array.isArray(value)) {
        throw new ValidationError(
          `context_schema_error: field '${field}' expected type array, got ${typeof value}`,
          SCHEMA_DOCS_URL,
        );
      }
      continue;
    }

    if (typeName === "object") {
      if (typeof value !== "object" || value === null || Array.isArray(value)) {
        const got = Array.isArray(value) ? "array" : value === null ? "null" : typeof value;
        throw new ValidationError(
          `context_schema_error: field '${field}' expected type object, got ${got}`,
          SCHEMA_DOCS_URL,
        );
      }
      continue;
    }

    // "string" | "boolean"
    if (typeof value !== typeName) {
      throw new ValidationError(
        `context_schema_error: field '${field}' expected type ${typeName}, got ${typeof value}`,
        SCHEMA_DOCS_URL,
      );
    }
  }
}

/** Return a new object with keys sorted. Stable order makes sign() deterministic. */
export function normalizeContext(
  context: Record<string, unknown> | null | undefined,
): Record<string, unknown> | undefined {
  if (context == null) return undefined;
  const sorted: Record<string, unknown> = {};
  for (const key of Object.keys(context).sort()) {
    sorted[key] = context[key];
  }
  return sorted;
}

/** Thrown when context fails the caller-supplied schema. */
export class ValidationError extends Error {
  /** Docs page for the structured-receipts feature. */
  docsUrl: string;

  constructor(message: string, docsUrl = SCHEMA_DOCS_URL) {
    super(message);
    this.name = "ValidationError";
    this.docsUrl = docsUrl;
  }
}
